### ISSUE TITLE
Display HTML representation even if other representations haven't finished loading.

### ISSUE THREAD

**Issue by theengineear** [ipython/ipython#7573]
For plotly, we've been returning display objects that define *many* representations that each require a separate request. When the user is online, IPython is always going to select the HTML representation. Is there anyway that this can be displayed immediately (before the other versions are downloaded?). It would seem that if HTML *will* top the other versions and it's already been received, IPython should just display it.

Is there a current way to do this?

**Comment by Carreau** [comment]
Hum, I'm not sure I completely get how you do it in your case. I *think* the mimebundle arrive at once on the frontend, but if it was arriving asynchronously, it would make sens.

**Comment by minrk** [comment]
A single display call results in a single message with all representations arriving simultaneously. I'm not sure what you are referring to regarding separate requests.

**Comment by theengineear** [comment]
Here's the code for our display object:
https://github.com/plotly/python-api/blob/master/plotly/tools.py#L1216-1263

When we return it on calls to make plots, we see the following happen:

(1) make a request to get the html
(2) wait to receive html
(3) for each static version:
    (a) make a request to get the image
    (b) wait for image to return
(4) display best pick (always html)

**Comment by minrk** [comment]
Ah, you are talking about requests in *your* code, not in ours. Now I get it.

Since the kernel doesn't know what format(s) will be displayed, all available representations are computed and sent to the frontend, which decides which to display. If you want to configure your own classes to have logic to selectively produce only one output under certain circumstances, you can do that by having the other repr methods return None.

**Comment by Carreau** [comment]
But I don't think they want to return `None`, they actually want all representation to be computed for nbconvert. 

I guess the best we could do is to fetch the representation asynchronously with a `yield from` or alike. 
I doubt we can right now, but maybe it is possible if you use futures in your init insted of each `res = requests.get(url)` so that the resources are fetched asynchronously, and then "just" need to be decoded when ipython inspect the differents `_repr_*_`.

**Comment by Carreau** [comment]
Note :  

(1) is due to you inhering our `IPython.core.display.HTML`, you don't need to
(3) is basically the core of our display machinery, we ask object to compute all their `_repr_*_`
once we have all (3) we send everythin to frontend, (here notebook) that choses the "right" representation, and store the rest along. We could also have a selector for user to select which repr they like.

**Comment by theengineear** [comment]
@Carreau , yes that's right, we *want* these things in the notebook for later. Async return of these things would be awesome for us as each call takes roughly the same amount of time.

This logic is on your end right now though, right? I.e., I can't do anything to get IPython to do these calls async.

What we're doing for now is just supporting fewer representations.

**Comment by Carreau** [comment]
> This logic is on your end right now though, right? I.e., I can't do anything to get IPython to do these calls async.

You can probably do something, as soon a one of the `_repr_`  is computed, you can trigger a methods that does fetch all the URL and use a cache on the instance.

**Comment by minrk** [comment]
Ah, you want mimebundles themselves to be fundamentally async. That's an interesting idea, but would require some pretty significant architectural changes in a variety of places.
